corehq/extensions/interface.py

```diff
@@ -0,0 +1,146 @@
+import importlib
+import inspect
+from collections import defaultdict
+
+import attr
+
+from dimagi.utils.logging import notify_exception
+from dimagi.utils.modules import to_function
+
+
+class ExtensionError(Exception):
+    pass
+
+
+@attr.s(frozen=True)
+class ExtensionPoint:
+    name = attr.ib()
+    providing_args = attr.ib()
+
+
+def extension_point(func):
+    """Decorator for creating an extension point"""
+    def extend(domains=None):
+
+        def _extend(impl):
```

Comment:
Would you agree it would be useful to have this raise an error if it is called after `corehq.extensions.app_config.ExtensionAppConfig.ready()`?

This could also add the extension to a list (something like `ExtensionAppConfig.extensions`), which would make it possible to eliminate the module traversal currently used to find extensions.

Reply:
I made an attempt at this here: fcd751a36cb but it's a bit messy

Reply:
I reworked it a bit to do what I had in mind. Feel free to cherry-pick https://github.com/dimagi/commcare-hq/commit/f623e8a1f5fc47027f65360154811e0947ee5d17 onto your branch if you like it (also feel free to modify it if you see issues).